src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -795,78 +796,152 @@ private string GetPreferredEnvironmentHost(string environmentHost, InstanceDisco
+            out AdalUsersForMsalResult adalUsersResult)
         {
-            var environment = new Uri(authority).Host;
             lock (LockObject)
             {
-                TokenCacheNotificationArgs args = new TokenCacheNotificationArgs
+                var args = new TokenCacheNotificationArgs
                 {
                     TokenCache = this,
                     ClientId = ClientId,
                     Account = null
                 };
 
-                IEnumerable<MsalRefreshTokenCacheItem> tokenCacheItems;
-                IEnumerable<MsalAccountCacheItem> accountCacheItems;
-                AdalUsersForMsalResult adalUsersResult;
-
                 OnBeforeAccess(args);
                 try
                 {
                     tokenCacheItems = ((ITokenCacheInternal)this).GetAllRefreshTokens(false);
                     accountCacheItems = ((ITokenCacheInternal)this).GetAllAccounts();
-                    adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(_logger, LegacyCachePersistence, ClientId);
+
+                    // Known Issue: will ignore env aliases, which we cannot get without an instance discovery
+                    adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(
+                        _logger,
+                        LegacyCachePersistence,
+                        ClientId,
+                        environment);
                 }
                 finally
                 {
                     OnAfterAccess(args);
                 }
+            }
+        }
 
-                IDictionary<string, Account> clientInfoToAccountMap = new Dictionary<string, Account>();
-                foreach (MsalRefreshTokenCacheItem rtItem in tokenCacheItems)
-                {
-                    foreach (MsalAccountCacheItem account in accountCacheItems)
-                    {
-                        if (rtItem.HomeAccountId.Equals(account.HomeAccountId, StringComparison.OrdinalIgnoreCase))
-                        {
-                            clientInfoToAccountMap[rtItem.HomeAccountId] = new Account
-                                (account.HomeAccountId, account.PreferredUsername, environment);
-                            break;
-                        }
-                    }
-                }
+        private async Task<string> GetPreferedCacheEnvAvoidNetworkCallAsync(
+            string authority,
+            IEnumerable<MsalAccountCacheItem> accountCacheItems)
+        {
+            // this map is to try to avoid a network call when getting accounts and it is ok if it is not up to date
+            // but it should not be used in any other scenario
+            IDictionary<string, string> envToPreferredCache = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+            {
+                { AzurePublicEnv, "login.windows.net" },
+                { "login.partner.microsoftonline.cn", "login.partner.microsoftonline.cn" },
```

Comment:
could you put these as consts somewhere, so if we have to change something, it's in one place?

Reply:
I'm hesitant to do this because we should only ever use these here. Not even in the file, but only in GetAccounts.